### PR TITLE
watch for devices via the flutter daemon

### PR DESCRIPTION
**Pull request by devoncarew**
Watch for devices via the flutter daemon.

- plumb in communication with the `flutter daemon` service
- listen for new devices, device state changes, and device removal

We still need to expose these devices to the user via some UI, and give them the ability to select which device an app will launch on.

@danrubel

### CODE REVIEW

File: lib/flutter/flutter_daemon.dart [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+  // TODO: result
+  // TODO: We need the stdout and stderr from launching the process.
+  Future<dynamic> start(
+    String projectDirectory, {
+    String target,
+    bool checked,
+    String route
+  }) {
+    return _call('app.start', _mapify({
```
Comment (danrubel) [inline lib/flutter/flutter_daemon.dart]: Why create a copy of the map? Is the map modified as part of the call?

File: lib/flutter/flutter_daemon.dart [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+class RequestError {
+  final String methodName;
+  final dynamic error;
+
+  RequestError(this.methodName, this.error);
+
+  String toString() => '[RequestError ${error}]';
+}
+
+Map _mapify(Map m) {
```
Comment (danrubel) [inline lib/flutter/flutter_daemon.dart]: Would it make sense to replace calls to this method with calls to Map.from(...) or does this method do something different that's important to its callers?

File: lib/flutter/flutter_daemon.dart [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+  FlutterSdkManager get _sdkManager => deps[FlutterSdkManager];
+
+  void _initSdk(FlutterSdk sdk) {
+    if (!_enableNotifications) return;
+
+    if (sdk == null) {
+      if (_daemon != null) {
+        _logger.info('Stopping Flutter daemon server');
+      }
+      _daemon?.dispose();
```
Comment (danrubel) [inline lib/flutter/flutter_daemon.dart]: Move this and the following line into the if (_daemon != null) block?

**Comment by danrubel** [comment]
LGTM

File: lib/flutter/flutter_daemon.dart (hunk 1 above)
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: I can rename this method, but what `_mapify` does is return a copy of the map with all keys stripped out that had associated `null` values. So a call site can create a map w/ all possible keys, and `_mapify` will strip out any keys that didn't have associated values.

File: lib/flutter/flutter_daemon.dart (hunk 2 above)
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: Different :) I'll rename to clarify it's behavior.

File: lib/flutter/flutter_daemon.dart (hunk 3 above)
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: done

File: lib/flutter/flutter_daemon.dart (hunk 2 above)
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: renamed to `_stripNullValues`

**Pull request merged**